pkg/user/external_server.go

```diff
@@ -0,0 +1,263 @@
+package user
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func readRequestBody(req *http.Request, body interface{}) error {
+	if err := json.NewDecoder(req.Body).Decode(body); err != nil {
+		return err
+	}
+	return nil
+}
+
+func sendResponse(rw http.ResponseWriter, status int, body interface{}) {
+	rw.Header().Set("Content-Type", "application/json")
+	rw.WriteHeader(status)
+
+	if body != nil {
+		// Set up the pipe to write data directly into the Reader.
+		pr, pw := io.Pipe()
+
+		// Write JSON-encoded data to the Writer end of the pipe.
+		// Write in a separate concurrent goroutine, and remember
+		// to Close the PipeWriter, to signal to the paired PipeReader
+		// that we’re done writing.
+		go func() {
+			pw.CloseWithError(json.NewEncoder(pw).Encode(body))
+		}()
+
+		_, err := io.Copy(rw, pr)
+		if err != nil {
+			log.WithError(err).Warn("[ExternalServer] error writing response")
+		}
+	}
+}
+
+func StartExternalServer(listenAddr string, stop <-chan struct{}, um UserManager) error {
+	ctx := context.Background()
+	s := &http.Server{
+		Addr: listenAddr,
+		Handler: http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			l := log.WithField("path", req.URL.Path).WithField("method", req.Method)
+			l.Debug("[ExternalUserManagerServer] request received")
+			switch fmt.Sprintf("%s %s", req.Method, req.URL.Path) {
```

Comment:
normally the practice is to use mux.Router, what is this?

Reply:
I started off with a mux.Router because that's what we'd done elsewhere, then threw it away in frustration and did this because it was much simpler in practice...

1. Just plain more lines of boilerplate to scroll through to actually see the code.
2. Having to register functions into the router outside of the Server.Handler function means I can't use lexical scoping to pass in the logger set up with `WithField` and either need to set it up from scratch in each handler or weave it in through a context variable, both of which are yet more unproductive lines obscuring the actual logic.

Although it's not too bad in a simple case like this, the mux.Router pattern means that control flow is alll over the place... when doing the authz refactoring in the gateway, I had to trace down all the middleware functions that every request was going through to work out what was really happening! The big list of `HandleFunc` calls in `server.go` that puts endpoint-specific middlewares on particular routes isn't so bad as you can see what's what there, but the global authentication/instrumentation middleware, plus the separate list hidden away in `admin_routes.go`... I'm still not sure I've not missed something somewhere in tracing the code that's actually executed to handle a request.

Reply:
but with mux.Router we could just grab it and register all these paths on the existing gateway HTTP server

Reply:
I don't plan to run it on the existing gateway HTTP server, putting it on a different port means it can be blocked off from public access at the network level (by not exposing that port publicly)!

Reply:
can we then have a normal HTTP server with router and testable handlers in the gateway and just remove this altogether?

Reply:
it would have been easier to use protobuf + grpc in this case if this is another HTTP server and not the one that we are already running, then you don't need to implement neither client nor server logic

Reply:
I'd vote in favour of consistency i.e using the standard `mux.Router`, as this code is going to worked on by other folks as well, and making use the the same patterns everywhere is a huge plus.